enterprise/coderd/identityprovider/tokens.go

```diff
@@ -0,0 +1,245 @@
+	// 40 characters matches the length of GitHub's client secrets.
+	rawRefreshToken, err := cryptorand.String(40)
+	if err != nil {
+		return oauth2.Token{}, err
+	}
+
+	// Generate the API key we will swap for the code.
+	// TODO: We are ignoring scopes for now.
+	tokenName := fmt.Sprintf("%s_%s_oauth_session_token", dbCode.UserID, app.ID)
+	key, sessionToken, err := apikey.Generate(apikey.CreateParams{
+		UserID:    dbCode.UserID,
+		LoginType: database.LoginTypeOAuth2ProviderApp,
+		// TODO: This is just the lifetime for api keys, maybe have its own config
+		//       settings. #11693
+		DefaultLifetime: defaultLifetime,
+		// For now, we allow only one token per app and user at a time.
+		TokenName: tokenName,
+	})
+	if err != nil {
+		return oauth2.Token{}, err
+	}
+
+	// Grab the user roles so we can perform the exchange as the user.
+	//nolint:gocritic // In the token exchange, there is no user actor.
+	roles, err := db.GetAuthorizationUserRoles(dbauthz.AsSystemRestricted(ctx), dbCode.UserID)
+	if err != nil {
+		return oauth2.Token{}, err
+	}
+	userSubj := rbac.Subject{
+		ID:     dbCode.UserID.String(),
+		Roles:  rbac.RoleNames(roles.Roles),
+		Groups: roles.Groups,
+		Scope:  rbac.ScopeAll,
+	}
+
+	// Do the actual token exchange in the database.
+	err = db.InTx(func(tx database.Store) error {
+		ctx := dbauthz.As(ctx, userSubj)
+		err = tx.DeleteOAuth2ProviderAppCodeByID(ctx, dbCode.ID)
+		if err != nil {
+			return xerrors.Errorf("delete oauth2 app code: %w", err)
+		}
+
+		// Delete the previous key, if any.
+		prevKey, err := tx.GetAPIKeyByName(ctx, database.GetAPIKeyByNameParams{
+			UserID:    dbCode.UserID,
+			TokenName: tokenName,
+		})
+		if err == nil {
+			err = tx.DeleteAPIKeyByID(ctx, prevKey.ID)
+		}
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return xerrors.Errorf("delete api key by name: %w", err)
+		}
+
+		newKey, err := tx.InsertAPIKey(ctx, key)
+		if err != nil {
+			return xerrors.Errorf("insert oauth2 access token: %w", err)
+		}
+
+		hashed := Hash(rawRefreshToken, app.ID)
+		_, err = tx.InsertOAuth2ProviderAppToken(ctx, database.InsertOAuth2ProviderAppTokenParams{
+			ID:           uuid.New(),
+			CreatedAt:    dbtime.Now(),
+			ExpiresAt:    key.ExpiresAt,
+			HashedSecret: hashed[:],
+			AppSecretID:  secret.ID,
+			APIKeyID:     newKey.ID,
+		})
+		if err != nil {
+			return xerrors.Errorf("insert oauth2 refresh token: %w", err)
+		}
+		return nil
+	}, nil)
+	if err != nil {
+		return oauth2.Token{}, err
+	}
+
+	return oauth2.Token{
+		AccessToken:  sessionToken,
+		TokenType:    "Bearer",
+		RefreshToken: rawRefreshToken,
+		Expiry:       key.ExpiresAt,
+	}, nil
+}
+
+/**
+ * Hash uses argon2 to hash the secret using the ID as the salt.
+ */
+func Hash(secret string, id uuid.UUID) []byte {
+	b := []byte(secret)
+	// TODO: Expose iterations, memory, and threads as configuration values?
+	return argon2.IDKey(b, []byte(id.String()), 1, 64*1024, 2, uint32(len(b)))
```

Comment:
I am using the app ID for this but I think the salt is supposed to be unique per secret.  I could prepend the ID of every secret to the secret, although that makes them pretty long.  Only the client secret is ever seen by a person though.

Reply:
I think we need to ditch this and use `userpassword.Hash`. I think it might be fine in practice, I just rather stick to the hash method we already know and use. I left another comment about how to handle the fact that `userpassword.Hash` is non-deterministic. 

Some notes on the security:
- `userpassword.Hash` is stronger because users might have the same password (`password123`) and these should be hashed to different values.
- This has function takes in a `secret` which already has a lot of randomness and is longer, so it does not need as strong of a hash function.

Still would rather use what we already use.

Reply:
> I am using the app ID for this but I think the salt is supposed to be unique per secret. 

For the best security yes. You could use `sha256(appID + userID)[:10]` or something deterministic but more unique. The best is a random salt, like done in `userpassword.Hash()`

> I could prepend the ID of every secret to the secret, although that makes them pretty long

We can make a new ID that is shorter. I put this comment above, but it just came to me if you want to use the `uuid` already and make it smaller, just do: `sha256(id)[:10]` and encode it with like base64 or something. The downside to just making a new ID for this purpose those is that base64 has is a non deterministic length encoding for any given input. (well you can predict it, but it is a pain).

So if we just choose random characters, then the ID is always a fixed length. We should not just use a `uuid` string though, as it is a terribly inefficient use of characters and is way to long.